Source/Scene/ModelExperimental/CustomShaderStage.js

```diff
@@ -0,0 +1,519 @@
+  // - _CUSTOM_ATTRIBUTE has an unknown type.
+  if (!defined(glslType)) {
+    return undefined;
+  }
+
+  return {
+    attributeField: generateStructField(glslType, attributeName),
+    value: value,
+  };
+}
+
+function generateVertexShaderLines(customShader, namedAttributes) {
+  var categories = partitionAttributes(
+    namedAttributes,
+    customShader._usedVariablesVertex.attributeSet
+  );
+  var addToShader = categories.addToShader;
+  var needsDefault = categories.missingAttributes;
+
+  var variableName;
+  var vertexInitialization;
+  var attributeFields = [];
+  var initializationLines = [];
+  for (variableName in addToShader) {
+    if (addToShader.hasOwnProperty(variableName)) {
+      var attribute = addToShader[variableName];
+      var attributeField = generateAttributeField(variableName, attribute);
+      attributeFields.push(attributeField);
+
+      // Initializing attribute structs are just a matter of copying the
+      // attribute or varying: E.g.:
+      // "    vsInput.attributes.position = a_position;"
+      vertexInitialization =
+        "    vsInput.attributes." + variableName + " = a_" + variableName + ";";
+      initializationLines.push(vertexInitialization);
+    }
+  }
+
+  for (var i = 0; i < needsDefault.length; i++) {
+    variableName = needsDefault[i];
+    var attributeDefaults = inferAttributeDefaults(variableName);
+    if (!defined(attributeDefaults)) {
+      // This primitive isn't compatible with the shader. To avoid compiling
+      // a shader with a syntax error, the custom shader will be disabled.
+      // this is indicated by returning an empty list of shader lines.
+      CustomShaderStage._oneTimeWarning(
+        "CustomShaderStage.incompatiblePrimitiveVS",
+        "Primitive is missing attribute " +
+          variableName +
+          ", disabling custom vertex shader"
+      );
+      return [];
+    }
+
+    attributeFields.push(attributeDefaults.attributeField);
+    vertexInitialization =
+      "    vsInput.attributes." +
+      variableName +
+      " = " +
+      attributeDefaults.value +
+      ";";
+    initializationLines.push(vertexInitialization);
+  }
+
+  // If there are no attributes, add a placeholder float field so the struct
+  // is not empty
+  if (attributeFields.length === 0) {
+    attributeFields.push("    float _padding;");
+  }
+
+  return [].concat(
+    "struct Attributes",
+    "{",
+    attributeFields,
+    "};",
+    "",
+    "struct VertexInput",
+    "{",
+    "    Attributes attributes;",
+    "};",
+    "",
+    "void initializeInputStruct(out VertexInput vsInput)",
+    "{",
+    initializationLines,
+    "}"
+  );
+}
+
+function generatePositionBuiltins(customShader) {
+  var fragmentInputFields = [];
+  var initializationLines = [];
+  var usedVariables = customShader._usedVariablesFragment.positionSet;
+
+  // Model space position is the same position as in the glTF accessor.
+  if ("positionMC" in usedVariables) {
```

Comment:
I think this is the first time I've seen `in` instead of `hasOwnProperty`. In this case there should be no difference because the object doesn't have inherited properties.

Reply:
Yeah I was wondering about if this was good style or not. I can switch to hasOwnProperty for consistency.